File: implementation/01 Fundamentals/Concurreny Basics/LRU Cache/go/lrucache.go
// lrucache.go
package lrucache

import (
	"fmt"
	"sync"
)

// node is an internal doubly linked list node.
// Not exported — implementation detail only.
type node[K comparable, V any] struct {
	key   K
	value V
	prev  *node[K, V]
	next  *node[K, V]
}

// LRUCache is a thread-safe, generic LRU cache.
// Zero value is not usable — must be created via New().
type LRUCache[K comparable, V any] struct {
	capacity int
	length   int
	cache    map[K]*node[K, V]
	head     *node[K, V] // sentinel MRU end
	tail     *node[K, V] // sentinel LRU end
	mu       sync.Mutex
}

// New creates a new LRUCache with the given capacity.
// Returns an error if capacity is not positive.
func New[K comparable, V any](capacity int) (*LRUCache[K, V], error) {
	if capacity <= 0 {
		return nil, fmt.Errorf("cache capacity must be positive, got %d", capacity)
	}

	head := &node[K, V]{}
	tail := &node[K, V]{}
	head.next = tail
	tail.prev = head

	return &LRUCache[K, V]{
		capacity: capacity,
		cache:    make(map[K]*node[K, V], capacity),
		head:     head,
		tail:     tail,
	}, nil
}

// Get retrieves a value by key. Returns the value and true if found,
// zero value and false if not. Marks the entry as most recently used.
func (c *LRUCache[K, V]) Get(key K) (V, bool) {
	c.mu.Lock()
	defer c.mu.Unlock()

	n, exists := c.cache[key]
	if !exists {
		var zero V
		return zero, false
	}
	c.moveToHead(n)
	return n.value, true
}

// Put inserts or updates a key-value pair.
// If at capacity, the least recently used entry is evicted.
func (c *LRUCache[K, V]) Put(key K, value V) {
	c.mu.Lock()
	defer c.mu.Unlock()

	if existing, exists := c.cache[key]; exists {
		existing.value = value
		c.moveToHead(existing)
		return
	}

	newNode := &node[K, V]{key: key, value: value}
	c.cache[key] = newNode
	c.insertAtHead(newNode)
	c.length++

	if c.length > c.capacity {
		lru := c.tail.prev
		c.removeNode(lru)
		delete(c.cache, lru.key)
		c.length--
	}
}

// Size returns the current number of entries in the cache.
func (c *LRUCache[K, V]) Size() int {
	c.mu.Lock()
	defer c.mu.Unlock()
	return c.length
}

// Clear removes all entries from the cache.
func (c *LRUCache[K, V]) Clear() {
	c.mu.Lock()
	defer c.mu.Unlock()
	c.cache = make(map[K]*node[K, V], c.capacity)
	c.head.next = c.tail
	c.tail.prev = c.head
	c.length = 0
}

// --- Private list operations ---
// Must only be called while holding mu.

func (c *LRUCache[K, V]) insertAtHead(n *node[K, V]) {
	n.prev = c.head
	n.next = c.head.next
	c.head.next.prev = n
	c.head.next = n
}

func (c *LRUCache[K, V]) removeNode(n *node[K, V]) {
	n.prev.next = n.next
	n.next.prev = n.prev
}

func (c *LRUCache[K, V]) moveToHead(n *node[K, V]) {
	c.removeNode(n)
	c.insertAtHead(n)
}

File: implementation/01 Fundamentals/Concurreny Basics/LRU Cache/java/LRUCache.java
// LRUCache.java
import java.util.HashMap;
import java.util.Map;
import java.util.concurrent.locks.ReentrantLock;

public class LRUCache<K, V> implements Cache<K, V> {

    private final int capacity;
    private final Map<K, Node<K, V>> cacheMap;
    private final Node<K, V> head; // sentinel — most recently used end
    private final Node<K, V> tail; // sentinel — least recently used end
    private final ReentrantLock lock;

    public LRUCache(int capacity) {
        if (capacity <= 0) {
            throw new IllegalArgumentException(
                "Cache capacity must be positive, got: " + capacity
            );
        }
        this.capacity = capacity;
        this.cacheMap = new HashMap<>(capacity);
        this.lock = new ReentrantLock();

        // Sentinel nodes eliminate null checks on every insert/remove.
        // head.next is always the MRU node.
        // tail.prev is always the LRU node.
        this.head = new Node<>(null, null);
        this.tail = new Node<>(null, null);
        head.next = tail;
        tail.prev = head;
    }

    @Override
    public V get(K key) {
        lock.lock();
        try {
            Node<K, V> node = cacheMap.get(key);
            if (node == null) {
                return null; // cache miss
            }
            moveToHead(node); // mark as most recently used
            return node.value;
        } finally {
            lock.unlock(); // always releases — even if exception thrown
        }
    }

    @Override
    public void put(K key, V value) {
        if (key == null) {
            throw new IllegalArgumentException("Cache key must not be null");
        }
        lock.lock();
        try {
            Node<K, V> existingNode = cacheMap.get(key);

            if (existingNode != null) {
                // Key exists — update value in-place, move to head.
                // No new node allocation, no eviction needed.
                existingNode.value = value;
                moveToHead(existingNode);
                return;
            }

            // New key — create node and insert at head (MRU position)
            Node<K, V> newNode = new Node<>(key, value);
            cacheMap.put(key, newNode);
            insertAtHead(newNode);

            // Evict LRU entry if over capacity.
            // Check + evict is atomic inside the lock — no race condition.
            if (cacheMap.size() > capacity) {
                Node<K, V> lruNode = tail.prev;
                removeNode(lruNode);
                cacheMap.remove(lruNode.key); // key stored on node for this
            }
        } finally {
            lock.unlock();
        }
    }

    @Override
    public int size() {
        lock.lock();
        try {
            return cacheMap.size();
        } finally {
            lock.unlock();
        }
    }

    @Override
    public void clear() {
        lock.lock();
        try {
            cacheMap.clear();
            head.next = tail;
            tail.prev = head;
        } finally {
            lock.unlock();
        }
    }

    // --- Private list operations ---
    // These are never called outside the lock. No synchronization needed here.

    private void insertAtHead(Node<K, V> node) {
        node.prev = head;
        node.next = head.next;
        head.next.prev = node;
        head.next = node;
    }

    private void removeNode(Node<K, V> node) {
        node.prev.next = node.next;
        node.next.prev = node.prev;
        // prev/next pointers left intact intentionally —
        // node is about to be GC'd, no need to null them out
    }

    private void moveToHead(Node<K, V> node) {
        removeNode(node);
        insertAtHead(node);
    }
}